app/src/main/java/org/thoughtcrime/securesms/jobs/MultiDeviceContactUpdateJob.java

```diff
@@ -0,0 +1,449 @@
+  {
+    if (length > 0) {
+      try {
+        SignalServiceAttachmentStream.Builder attachmentStream = SignalServiceAttachment.newStreamBuilder()
+                                                                                        .withStream(stream)
+                                                                                        .withContentType("application/octet-stream")
+                                                                                        .withLength(length)
+                                                                                        .withResumableUploadSpec(messageSender.getResumableUploadSpec());
+
+        messageSender.sendSyncMessage(SignalServiceSyncMessage.forContacts(new ContactsMessage(attachmentStream.build(), complete)),
+                                      UnidentifiedAccessUtil.getAccessForSync(context));
+      } catch (IOException ioe) {
+        throw new NetworkException(ioe);
+      }
+    } else {
+      Log.w(TAG, "Nothing to write!");
+    }
+  }
+
+  private Optional<SignalServiceAttachmentStream> getAvatar(@NonNull RecipientId recipientId, @Nullable Uri uri) {
+    Optional<SignalServiceAttachmentStream> stream;
+
+    if (SignalStore.settings().isPreferSystemContactPhotos()) {
+      stream = getSystemAvatar(uri);
+
+      if (!stream.isPresent()) {
+        stream = getProfileAvatar(recipientId);
+      }
+    } else {
+      stream = getProfileAvatar(recipientId);
+
+      if (!stream.isPresent()) {
+        stream = getSystemAvatar(uri);
+      }
+    }
+
+    return stream;
+  }
+
+  private Optional<SignalServiceAttachmentStream> getProfileAvatar(@NonNull RecipientId recipientId) {
+    if (AvatarHelper.hasAvatar(context, recipientId)) {
+      try {
+        long length = AvatarHelper.getAvatarLength(context, recipientId);
+        return Optional.of(SignalServiceAttachmentStream.newStreamBuilder()
+                                                        .withStream(AvatarHelper.getAvatar(context, recipientId))
+                                                        .withContentType("image/*")
+                                                        .withLength(length)
+                                                        .build());
+      } catch (IOException e) {
+        Log.w(TAG, "Failed to read profile avatar!", e);
+        return Optional.absent();
+      }
+    }
+
+    return Optional.absent();
+  }
+
+  private Optional<SignalServiceAttachmentStream> getSystemAvatar(@Nullable Uri uri) {
+    if (uri == null) {
+      return Optional.absent();
+    }
+
+    if (!Permissions.hasAny(context, Manifest.permission.READ_CONTACTS, Manifest.permission.WRITE_CONTACTS)) {
+      return Optional.absent();
+    }
+
+    Uri displayPhotoUri = Uri.withAppendedPath(uri, ContactsContract.Contacts.Photo.DISPLAY_PHOTO);
+
+    try {
+      AssetFileDescriptor fd = context.getContentResolver().openAssetFileDescriptor(displayPhotoUri, "r");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *fd*. Currently, there are execution paths that do not contain closure statements, for example, when AssetFileDescriptor.createInputStream throws an exception. Ensure *fd* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).